ocaml/otherlibs/systhreads/st_stubs.c

```diff
@@ -771,21 +834,34 @@ CAMLprim value caml_thread_exit(value unit)   /* ML */
 
 CAMLprim value caml_thread_yield(value unit)        /* ML */
 {
-  if (st_masterlock_waiters(&caml_master_lock) == 0) return Val_unit;
-
-  /* Do all the parts of a blocking section enter/leave except lock
-     manipulation, which we'll do more efficiently in st_thread_yield. (Since
-     our blocking section doesn't contain anything interesting, don't bother
-     with saving errno.)
-  */
-  caml_raise_async_if_exception(caml_process_pending_signals_exn(),
-                                "signal handler");
-  caml_thread_save_runtime_state();
-  st_thread_yield(&caml_master_lock);
-  curr_thread = st_tls_get(thread_descriptor_key);
-  caml_thread_restore_runtime_state();
-  caml_raise_async_if_exception(caml_process_pending_signals_exn(),
-                                "signal handler");
+  struct caml_locking_scheme* s;
+
+  s = atomic_load(&caml_locking_scheme);
+  if (s->count_waiters != NULL && s->yield != NULL) {
+    if (s->count_waiters(s->context) == 0) return Val_unit;
```

Comment:
Why not have `can_skip_yield` that returns a bool instead of `count_waiters`? Then it can be implemented by all locking schemes.

Reply:
I think can_skip_yield is a better API. What do you mean by "it can be implemented by all locking schemes"? (Maybe you mean that there's a reasonable no-op behaviour there? If you mean that all locking schemes can detect other waiters, then I don't follow)

Reply:
I mean there is a reasonable no-op behaviour. Any implementation can safely return `false`, whereas not every implementation can return an accurate number of waiters.